asc2cdf/asc2cdf.man
.na
.nh
.TH ASC2CDF 1 "12 March 1998" "Local Command"
.SH NAME
asc2cdf \- ASCII to netCDF converter
.SH SYNOPSIS
asc2cdf [-b time_t] [-a] [-l] [-m] [-r n] [-s n] ASCII_file netCDF_file
.SH DESCRIPTION
.B asc2cdf 
is program for uploading ASCII data into a netCDF file that follows the
NCAR-RAF/nimbus conventions.  Files that are then useable by ncplot and
the suite of nc* programs at RAF.  With no options, asc2cdf expects the
file to look like something produced by n2asc.  A top row with variable
names, and time in the HH:MM:SS format in the left column.
.PP
The -a option tells asc2cdf that the input file will be following the NASA
Data Exchange Format (Ames).
.PP
The -l option tells asc2cdf that the input file will be following the NASA
Data Exchange Format (Langley).
.PP
The -b option defines a base_time in Unix seconds since January 1st 1970. 
This is useful when trying to merge data back into a netCDF file, where
ncmerge(1) complains about no overlapping time segments.  The default value
is 0.  This option has no affect when the -a or -l option is enabled.
.PP
The -r option defines the data rate.  Where the rate must be less than 1 Hz.
n = 5 means 1 sample every 5 seconds.  The default is 1.
.PP
The -s option tells asc2cdf to skip 'n' lines before getting to the data.
This option is ignored if -a or -l option has been specified.  The default is 1.
.PP
The -m option tells asc2cdf that time in the left column will be in
seconds since midnight format (with rounding).  This is automatically
turned on with the -a or -l option.
.PP
.SH AUTHOR
Chris Webster, National Center for Atmospheric Research


doc/HISTORY

Version number is changed in XmNimbus.
-------------------------------------------------------------------------
11/30/98	2.5.5
	MASP support.  Cell sizes are end-points now, not mid-points.
	260x changes/fixes, including removal physical bin-shift between
	ACCUM & CONC, now just adjust starting.

05/19/98	2.5.4
	netCDF mods, removed global defaults, and stored default with given
	variable.  Made nimbus check default directory for Defaults
	then local proj dir.  netCDF file version 1.1.  fixed fssp range
	affecting HRT sampled FRNG.  Cleaned up all processing of HRT
	sampled PMS1D (mostly housekeeping).  Corrections to 260x for
	HRT sampling also.  Modified to handle 30 minute time gaps before
	inserting new time period.  New AKRD calcoes for C130.

01/02/98	2.5.3
	netCDF mods.  PMSspecs recorded with AFSSP, etc.  Only Dimensions
	used are created.  Categories added to netCDF file.  Solar angle,
	radiation corrections.

10/01/97	2.5.2
	DAY, MONTH, YEAR added.  Fixed Production output file name, if
	FlightNum was changed.  Log File name corrected for LRT/HRT from
	setup file & if flighNum changed.

05/02/97	2.5.0
	RealTime mode added.  CAINE-II variables.

04/02/97	2.4.2
	Fixed bugs in Toggle Probe.  Tacked h onto HRT production file name.
	ASTG for PALT moved to Defaults.  Fluxes off by default.  Added
	PSURF_ALT_MAX to defaults.

11/06/96	2.4.0
	WP3 modified.  Support for new APN159.  PMSspecs file.  New high rate
	filters.  Fixed minor discrepancies in AMLIB.  Goff-Gratch for EDPC.
	Transport delays for IRS.  Dynamic lags for GPS.  Option to disable
	all lagging && despiking.

07/31/96	2.3.2
	DBZ calc fix.  Support for ex? dropped.  Production log file.
	(Supposedly) Improved file name validation.  Despike stats.  More
	B-57 AMLIB.  AKRD finalized for C-130.  Read COS blocked files.

04/14/96	2.3.1
	Differences.  New IRS parms, TKA, DRFTA, TKAR.  GIV.  B57.
	Condensed all refernce fns to one.  RATIO.  DropSonde.  Collins GPS.
	Modified AKRD for C-130.  EVTCKP renamed to EVENT_CKP.

02/11/96	2.3.0
	Fixed bug in SumCounter (was truncating), bug only affected Counters
	with calcos.  When nimbus asked for flight number, it then failed to
	record it in netCDF file.  ACE variable XFCN added. GIV.  Filled data
	gaps had time stamp of -32767:-32767:-32767, fixed.  Added 'Edit'
	Menu option to edit flight info (Date, Flt#, etc).  Added
	Dismiss/Reset to TimeSlice window.

10/23/95	2.2.1
	Added support for high-rate processing of counters sampled at 1 hz.
	Fixed bug in PSFD & INPWANT.  Flying in circles.  Allowed cal coes
	for counters.

09/15/95	2.2.0
	Fixed CheckForTimeGap to handle all time malfunctions.  260x dof alg
	change.  Added more chemistry vars for ACE.  Added spike slope to
	setup files & Edit window.  Async data file/storage added.
	Reflectivity added for FSSP-100 & ????.  Community inlet pressure
	differential variable added.  Mod PCAS code to use FLOW for Sample
	Volume.

08/13/95	2.1.6
	Another despike bug fixed.  Checks for multiple ADS files, and
	crosses files automatically (mass store limitation).  C130 Pcor
	corrections.

06/06/95	2.1.5
	Despike bug & Sailplane bug fixed.

05/12/95	2.1.4
	Finished Sailplane stuff.

04/27/95	2.1.3
	Hard code output filename for production runs.  New code for TASHC.
	Community Inlet parameters added.

04/13/95	2.1.2
	Bug fixes, corrected winds hr bug, added fluxes to toggle probe.
	Fixed Toggle Rate bug.  X's put in front of Corrected WINDS.  TASHC.
	Sailplane stuff.  All probe's limited to 16.

??/??/95	2.1.1 - ?????

02/01/95	2.1.0
	Toggle Probe menu item. C130 PCORS.  Fixed XSO2.  New 250hz decimating
	filter (5:1 only).  Don't HR variables that don't need it.  added
	convertOffset.  New method of decoding IRS data.

12/22/94	2.0.1
	Some bug fixes, Coeff for WP3 barrel modified and fixed all Aircraft
	WP3.

12/19/94	2.0
	Averaging modified, to apply cal coes after averaging for real-time,
	saving 10-15% in procesing of Analog channels.  Overhauled access to
	data in amlib, and modified HighRate data storage, causing 10-15%
	decrease in processing time of high rate runs.  ^R added for toggling
	output rates.  Converted to ANSI C.  Fixed bugs in batch processing.
	Despiking added.  SampleRate computation of PMS1D derived varibles.
	Fixed constants in PSURF calculation.

11/09/94	1.8
	Fixed high-rate for counter variables.  PALTF.  WIC bug fixed.
	FPCRC bug fixed.  DependedUpon, used for removing unwanted variables
	from computation.

08/31/94	1.7
	support for 1000 hz.  New PSFD code for DSM.  Disk full fix, Warning
	on non-production files.  Re-arranged some PMS1D stuff, by adding
	fake1d.c.  More C130 support.

07/15/94	1.6
	HI3 -> ALT, C130 support, base_time bug, PMS1D bin Output defaults
	TRUE, Mark Dirty vars for user to see, created file for LowRateVars.
	Bug fixes.  FLUXes to AMLIB.

04/??/94	1.5
	hr/batch bug.  New hr decimating filters.

02/??/94	1.4
	added TimeSlice time verification.  EditWindow, add full variable
	name.  More bug fixes.

02/15/94	1.3
	Removed ADS header text field.

02/10/94	1.2
	added derived variables ZERO, ONE to ALWAYS.  Straightened out the
	boys in Chemistry.

01/26/94	1.1
	released, bug fixes, changes to CVI, DPCRC, added FastOzone.

11/25/93	Nimbus 1.0 released.


nc_utils/asc2cdf_HRT/asc2cdf.c
/*
-------------------------------------------------------------------------
OBJECT NAME:	asc2cdf.c

FULL NAME:	ASCII to Nimbus-netCDF Low Rate

ENTRY POINTS:	main()

STATIC FNS:		

DESCRIPTION:	Translate ASCII file to Nimbus Low Rate netCDF file

COPYRIGHT:	University Corporation for Atmospheric Research, 1996-7
-------------------------------------------------------------------------
*/

#define _XOPEN_SOURCE
#include "define.h"
#include <raf/constants.h>

#include <sys/types.h>
#include <sys/stat.h>

char	buffer[BUFFSIZE];

int	ncid;
int	baseTimeID, timeOffsetID, timeVarID, varid[MAX_VARS],
	nVariables;
time_t	BaseTime = 0;
float	scale[MAX_VARS], offset[MAX_VARS], missingVals[MAX_VARS];

char	*time_vars[] = {"HOUR", "MINUTE", "SECOND", 0 };
struct tm StartFlight;

static FILE     *inFP;
static size_t   nRecords;
static char     *globalAttrFile = 0;

/* Command line option flags.
 */
bool	fileType = PLAIN_FILE, sexSinceMidnight = false;
int	SkipNlines = 1;

int		BaseDataRate = 1, dataRate = 10;
const int	rateOne = 1;

const char	*noTitle = "No Title";
const char	*noUnits = "Unk";


static int ProcessArgv(int argc, char **argv);
static void WriteMissingData(int, int);


/* -------------------------------------------------------------------- */
int main(int argc, char *argv[])
{
  int	i, hz, hour, minute, second, currSecond;
  int	startHour, startMinute, startSecond;
  char	*p;
  float	dataValue;
  size_t	index[2];

  i = ProcessArgv(argc, argv);

  if ((argc - i) < 2)
    {
    fprintf(stderr, "Usage: asc2cdf [-a] [-l] [-m] [-r n] [-s n] ascii_file output.nc\n");
    exit(1);
    }

  if ((inFP = fopen(argv[i], "r")) == NULL)
    {
    fprintf(stderr, "Can't open %s.\n", argv[i]);
    exit(1);
    }


  if (nc_create(argv[i+1], NC_CLOBBER, &ncid) != NC_NOERR)
    {
    fprintf(stderr, "Can't destroy %s.\n", argv[i+1]);
    exit(1);
    }


  switch (fileType)
    {
    case PLAIN_FILE:
      CreatePlainNetCDF(inFP);
      SetPlainBaseTime();
      break;

    case NASA_AMES:
      CreateNASAamesNetCDF(inFP);
      SetNASABaseTime();
      break;

    case NASA_LANGLEY:
      CreateNASAlangNetCDF(inFP);
      SetNASABaseTime();
      break;
    }


  if (BaseDataRate < 1)
    BaseDataRate = 1;

  printf("Averaging Period = %d, Data Rate = %dHz\n", BaseDataRate, dataRate);


  /* Start uploading data.
   */
  rewind(inFP);

  while (SkipNlines--)
    fgets(buffer, BUFFSIZE, inFP);


  for (nRecords = 0; fgets(buffer, BUFFSIZE, inFP); )
    {
    if (strlen(buffer) < 2)
      continue;

    p = strtok(buffer, ", \t");

    if (fileType == NASA_LANGLEY)	/* Skip Julian day	*/
      p = strtok(NULL, ", \t");


    if (sexSinceMidnight)
      {
//      currSecond = atof(p);

      if (nRecords == 0)	// Roll our own time, for non-uniform time.
        currSecond = atoi(p);
      else
        ++currSecond;

//if (atoi(p) == 70200) printf("19:30 = %d\n", currSecond);
//if (atoi(p) == 73800) printf("20:30 = %d\n", currSecond);

      hour = currSecond / 3600; currSecond -= hour * 3600;
      minute = currSecond / 60; currSecond -= minute * 60;
      second = currSecond;
      }
    else
      {
      hour = atoi(p);
      minute = atoi(&p[3]);
      second = atoi(&p[6]);
      }

    if (hour > 23)
      hour -= 24;

    currSecond = hour * 3600 + minute * 60 + second;

    if (nRecords == 0)
      {
      startHour = hour;
      startMinute = minute;
      startSecond = second;
      }

    dataValue = (float)(nRecords * BaseDataRate);
    nc_put_var1_float(ncid, timeOffsetID, &nRecords, &dataValue);
    nc_put_var1_float(ncid, timeVarID, &nRecords, &dataValue);
    dataValue = (float)hour;
    nc_put_var1_float(ncid, varid[0], &nRecords, &dataValue);
    dataValue = (float)minute;
    nc_put_var1_float(ncid, varid[1], &nRecords, &dataValue);
    dataValue = (float)second;
    nc_put_var1_float(ncid, varid[2], &nRecords, &dataValue);

    for (hz = 0; hz < dataRate; ++hz)
      {
      for (i = 0; i < nVariables; ++i)
        {
        p = strtok(NULL, ", \t\r\n");
        dataValue = atof(p);

        if (fileType != PLAIN_FILE)
          {
          if (dataValue == missingVals[i])
            dataValue = MISSING_VALUE;
          else
            dataValue = dataValue * scale[i] + offset[i];
          }

        index[0] = nRecords; index[1] = hz;
        nc_put_var1_float(ncid, varid[i+3], index, &dataValue);
        }

      if (hz != dataRate-1)
        if (fgets(buffer, BUFFSIZE, inFP) == NULL)
          break;

      p = strtok(buffer, ", \t");

      if (fileType == NASA_LANGLEY)	/* Skip Julian day	*/
        p = strtok(NULL, ", \t");
      }

    ++nRecords;
    }


  sprintf(buffer, "%02d:%02d:%02d-%02d:%02d:%02d",
          startHour, startMinute, startSecond, hour, minute, second);
printf("TimeInterval = %s\n", buffer);
  nc_put_att_text(ncid, NC_GLOBAL, "TimeInterval", strlen(buffer)+1, buffer);
  nc_close(ncid);
  chmod(argv[2], 0666);

  fclose(inFP);

  return(0);

}	/* END MAIN */

/* -------------------------------------------------------------------- */
void AddTimeVariables(int dims[])
{
  nc_def_var(ncid, "base_time", NC_INT, 0, 0, &baseTimeID);
  strcpy(buffer, "seconds since 1970-01-01 00:00:00 +0000");
  nc_put_att_text(ncid, baseTimeID, "units", strlen(buffer)+1, buffer);
  strcpy(buffer, "Start time of data recording.");
  nc_put_att_text(ncid, baseTimeID, "long_name", strlen(buffer)+1, buffer);

  nc_def_var(ncid, "Time", NC_INT, 1, dims, &timeVarID);
  strcpy(buffer, "time of measurement");
  nc_put_att_text(ncid, timeVarID, "long_name", strlen(buffer)+1, buffer);
  strcpy(buffer, "time");
  nc_put_att_text(ncid, timeVarID, "standard_name", strlen(buffer)+1, buffer);

  nc_def_var(ncid, "time_offset", NC_FLOAT, 1, dims, &timeOffsetID);
  strcpy(buffer, "Seconds since base_time.");
  nc_put_att_text(ncid, timeOffsetID, "long_name", strlen(buffer)+1, buffer);


  strftime(buffer, 256, "seconds since %F %T +0000", &StartFlight);
  nc_put_att_text(ncid, timeVarID, "units", strlen(buffer)+1, buffer);
  nc_put_att_text(ncid, timeOffsetID, "units", strlen(buffer)+1, buffer);

}

/* -------------------------------------------------------------------- */
static void WriteMissingData(int currSecond, int lastSecond)
{
  int	i, j, ts;
  int	hour, minute, second;
  float	dataValue;

  ts = (lastSecond += BaseDataRate);

  hour = ts / 3600; ts -= hour * 3600;
  minute = ts / 60; ts -= minute * 60;
  second = ts;

  for (i = lastSecond; i < currSecond; i += BaseDataRate, ++nRecords)
    {
    dataValue = (float)(nRecords * BaseDataRate);
    nc_put_var1_float(ncid, timeOffsetID, &nRecords, &dataValue);
    dataValue = (float)hour;
    nc_put_var1_float(ncid, varid[0], &nRecords, &dataValue);
    dataValue = (float)minute;
    nc_put_var1_float(ncid, varid[1], &nRecords, &dataValue);
    dataValue = (float)second;
    nc_put_var1_float(ncid, varid[2], &nRecords, &dataValue);

    dataValue = MISSING_VALUE;

    for (j = 0; j < nVariables; ++j)
      nc_put_var1_float(ncid, varid[j+3], &nRecords, &dataValue);

    if ((second += BaseDataRate) > 59)
      {
      second = 0;
      if (++minute > 59)
        {
        minute = 0;
        if (++hour > 23)
          hour = 0;
        }
      }
    }

}	/* END WRITEMISSINGDATA */

/* -------------------------------------------------------------------- */
static int ProcessArgv(int argc, char **argv)
{
  int	i;

  for (i = 1; i < argc; ++i)
    {
    if (argv[i][0] != '-')
      break;

    switch (argv[i][1])
      {
      case 'b':
        strptime(argv[++i], "%F %T +0000", &StartFlight);
        BaseTime = timegm(&StartFlight);
//        BaseTime = mktime(&StartFlight) - timezone;
        break;

      case 'm':
        sexSinceMidnight = true;
        break;

      case 'a':
        fileType = NASA_AMES;
        sexSinceMidnight = true;
        break;

      case 'l':
        fileType = NASA_LANGLEY;
        sexSinceMidnight = true;
        break;

      case 'r':
        BaseDataRate = atoi(argv[++i]);
        break;

      case 's':
        SkipNlines = atoi(argv[++i]);
        break;

      default:
        fprintf(stderr, "Invalid option %s, ignoring.\n", argv[i]);
      }

    }

  return(i);

}	/* END PROCESSARGV */

/* END ASC2CDF.C */
